commons/src/config/preamble.rs

```diff
@@ -60,12 +76,156 @@ impl ConfigFns for I18n {
     }
 
     fn validate(&self) -> Result<(), ConfigErr> {
-        // TODO: make sure strings are valid bcp-47 locales
+        let mut locales: Vec<_> = std::iter::once(&self.lang)
+            .chain(self.langs.iter())
+            .collect();
+
+        locales.sort_by_key(|locale| locale.to_string());
+        locales.dedup();
+
+        if self.locales_file.is_none() {
+            let allowed_locales = [
+                langid!("en"),
+                langid!("en-US"),
+                langid!("en-UK"),
+                langid!("de"),
+                langid!("de-DE"),
+                langid!("de-AT"),
+                langid!("bs"),
+                langid!("bs-BA"),
+            ];
+
+            if !locales
+                .iter()
+                .all(|langid| allowed_locales.contains(&langid.id))
+            {
+                return err!(ConfigErr::BadLocaleUsed);
+            }
+        }
+
+        let blob = self.get_blob()?;
+
+        // check if it loads
+        let provider =
+            icu_provider_blob::BlobDataProvider::try_new_from_blob(blob.into_boxed_slice())
+                .map_err(|_| {
+                    pipe!(
+                        ConfigErr::InvalidFile,
+                        &format!(
+                            "Failed to read locales file: {}",
+                            self.locales_file.as_ref().unwrap().to_string_lossy()
+                        )
+                    )
+                })?;
+
+        let key = icu_datagen::key("decimal/symbols@1").expect("decimal/symbols@1 is a valid key.");
+        for locale in locales {
+            let req = DataRequest {
+                locale: &(locale).into(),
+                metadata: Default::default(),
+            };
+
+            if provider.load_buffer(key, req).is_err() {
+                logid::log!(
+                    ConfigErr::LocaleMissingKeys,
+                    add: AddonKind::Info(
+                        format!(
+                            "Locale {} is not contained in the provided file. Using fallback locale '{}'.",
```

Comment:
```suggestion
                            "Locale {} is not contained in the provided file. Trying to use fallback locale '{}'.",
```

Slight rephrase of info msg, since the fallback locale might still fail below.